Nonprofit Cloud Integration

gabrielcouto357 opened the pull request:
Updates to include ability to map to NPC objects GiftCommitment, GiftTransaction and Person Accounts

gabrielcouto357 committed:
Object Updates

Added new types for frField__mdt and NPC_Setting__c is used to reference NPC fields without having to reference it

gabrielcouto357 committed:
Apex class updates

Updated Apex classes to check if its a NPC org, if the object exists instead maps to its NPC counterpart

gabrielcouto357 committed:
Update FrSetup VF Page

Added ability to make mapping for new NPC objects, including a stop to make sure fr_id__c is available on each object (GiftTransaction, GiftCommitment, and Account)

gabrielcouto357 committed:
Added new Metadata

The mapping for the id from request, needed for all NPC Objects

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -62,76 +62,180 @@ public class frDonation extends frModel implements frSyncable {
+        if (Schema.getGlobalDescribe().containsKey('gifttransaction')) {
+            //add a check for a dupe transaction?
+            String objectName = 'gifttransaction';  
+            List<SObject> gtList = Database.query('SELECT Id, Name FROM GiftTransaction WHERE fr_ID__c = :frId');
```
How will we be able to create the fr_ID__C field on the GiftTransaction object?  

Can we include that custom field in this repository & will that still be valid to deploy to organizations that do not have a GiftTransaction object?

slpeterson replied:
The idea was not put any fields or objects relating to NPC objects as in the code otherwise it'll be a package dependency.  That field would need to be created and mapped post install.  We added validation on the mapping screens that force a field with the API fr_ID__c to exist and to be mapped to allow a successful save.  There are probably some ways to make the setup more elegant but that was the intent.

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -62,76 +62,180 @@ public class frDonation extends frModel implements frSyncable {
+            Boolean alreadyMappedToPledge = alreadyMappedToPledgeList.size() > 0;
+            if(!isPledge && !alreadyMappedToPledge) {
+                Pledge__c pledge = frPledge.findActive((Id) gt.get('DonorId'), (Date) gt.get('AcknowledgementDate'));
+                NPC_Setting__c npcSetting = NPC_Setting__c.getOrgDefaults();
```
Can you explain more on the purpose of this new custom setting that seems to be storing a global Pledge value?  I think it would be unexpected for every transaction to count towards the same pledge

gabrielcouto357 replied:
The custom setting value holds the name of the field to populate for pledge since we cant create the lookup due to it being an NPC object

Alexalex1432 replied:
Hm this would mean that in NPC orgs we'll have to instruct them that a custom lookup field and custom setting needs to be created and then they will have to enter that in the custom setting?

Can we change this so that the field name will be something that we hardcode, and log a funraise_Error__c if they have not created the lookup?

gabrielcouto357 replied:
Sorry didnt respond to this one but this has been updated in a previous update!

Alexalex1432 commented:
As a general note, can we remove any System.debug or commented out code that won't be needed.  If there are any comments that you need input from me on, please tag in a comment here

Alexalex1432 commented:
We will need to update the `getFields()` and `getObjects()` method in any class that extends frModel that were changed here, as these are used for checking permissions and that was a required change from Salesforce as part of our App Security review, that we need to ensure we have permissions for the objects/fields that will be modified.

Alexalex1432 commented:
in frUtil we have a method called `hasNPSP()` to check if the organization has the nonprofit success pack.  Can we refactor the current schema checks for NPC functionality to a similar centralized method so it can be more easily managed and clearly show we are entering a blocks of logic for NPC

Alexalex1432 commented:
`getSalesforceId()` should be updated to return the gift transaction ID in the NPC situations, as this will be used in various places, including field mappings that are queried for in `frModel.applyMappings`  which are used to pull the current record and resolve data conflicts between Funraise & Salesforce

Alexalex1432 commented:
Discussed elsewhere, but I'll also leave a comment here to keep track that we need to add test coverage for this new logic and ensuring a sync from Funraise results in the correct records and data being created (along with providing the required coverage enforced by Salesforce)

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -62,76 +62,180 @@ public class frDonation extends frModel implements frSyncable {
+            gt.TransactionDueDate = System.today();*/
+            applyMappings(gt, request);
+            System.debug('out of mappings');
+            //setOpportunityStage(o, String.valueOf(request.get('status'))); TODO: Have a similar function but for gt.Status
```
Leaving a comment here to keep track of this TODO for mapping Funraise Status -> GT Status

Alexalex1432 commented on src/classes/frSetupController.cls:
```diff
@@ -5,6 +5,27 @@ public with sharing class frSetupController {
 	}
+	public static String PERSON_ACCOUNT_TYPE {
+		get {
+			//TODO Put it in the class, maybe a new field on frField to diff between the 2?
```
Leaving a comment here to keep track of this TODO for refactoring the types (if we decide to)

Alexalex1432 commented on src/classes/frSetupController.cls:
```diff
@@ -90,7 +178,7 @@ public with sharing class frSetupController {
 		}
 		donationSFOptions.sort();
-		for(frField__mdt field : [SELECT DeveloperName, MasterLabel FROM frField__mdt WHERE Type__c = :DONATION_TYPE ORDER BY MasterLabel]) {
+		for(frField__mdt field : [SELECT DeveloperName, MasterLabel FROM frField__mdt WHERE Type__c = :DONATION_TYPE OR Type__c = 'All' ORDER BY MasterLabel]) {
```
How does a field of Type ALL work here?  Generally, the field definitions are managed per the type of entities that sync and are mappable from Funraise, is there a type of field coming from Funraise that would exist on all mappable entities?

Alexalex1432 replied:
I wanted to follow up on this one as I'm not following the reasoning for a field having a type__c of ALL

gabrielcouto357 replied:
The type ALL is just used to get the funraise id to populate as an option to map somewhere in the frSetup page mainly to validate that the fields exist during setup

Alexalex1432 commented on src/classes/frSetupController.cls:
```diff
@@ -157,20 +268,51 @@ public with sharing class frSetupController {
+				addConfirm('Mappings saved successfully');
+			}
+		} else {
+			addError('fr_ID__c Field must be mapped and present on all NPC Objects.');
```
I think we don't want to enforce that users map the Funraise ID field themselves, as since we know it is required we should handle mapping ourselves in the logic that saves the records, similar to how it works today for Opportunites & Contacts

Can we replace this with validation that the field exists (since, as we previously discussed we won't be able to create the custom fields on the NPC objects, we'll need a way to ensure the custom fields we need are there)

Alexalex1432 replied:
I want to follow up on this one as well, let me know any thoughts!

gabrielcouto357 replied:
We could do this but im worried since we cant create the fields with the package that without a way in setup to warn/notify the user that the fields dont exist it could create confusion when the package isnt working as expected. 

I could replace it with validation that the field exists and have a warning banner on the setup page if they dont, would this be better?

Alexalex1432 replied:
I'm happy with validation that the field exists, but once it exists we should not be forcing the user to map to values it

gabrielcouto357 replied:
Sorry didnt respond to this one but this has been updated in a previous update, was decided that a post deployment step would be sufficient!

gabrielcouto357 committed:
Update FrSetup VF Page

gabrielcouto357 committed:
Update FrSetup VF Page

gabrielcouto357 committed:
Updated Apex classes based on feedback

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +55,126 @@ public class frDonation extends frModel implements frSyncable {
     
-    private Opportunity o;
+    private SObject o;
+    private SObject gt;
```
It looks like the `o` object has been made generic and is used in both branches of NPC/NPSP, do we need this GT object in that case?

gabrielcouto357 replied:
Nope not needed! removed in new update.

Alexalex1432 replied:
It looks like this field wasn't removed, can you double check if the most recent commit was sent properly?

gabrielcouto357 replied:
Seems to be removed by a previous update, could be the git window showing here being outdated?

Alexalex1432 commented on src/classes/frModel.cls:
```diff
@@ -214,10 +217,7 @@ public with sharing abstract class frModel {
-                                +' Record Id: '+record.Id+' - Funraise Id: '+ funraiseId + ' - Field: '+field.getDescribe().getName()+' - Value: '+value
-                                +' Exception: '+ex
-                               );
+            System.debug('field: ' + ex);
```
Can we restore the error logging here and any other places it was replaced with debugging?

gabrielcouto357 replied:
Fixed here and a couple of other places as well, let me know if i missed any and i'll get right on it ty!

Alexalex1432 commented on src/classes/frSetupController.cls:
```diff
@@ -78,8 +109,58 @@ public with sharing class frSetupController {
+        }
+		//Check for GiftTransaction to see if its a NPC Org
+		if (frUtil.hasNPCobjects()) {
+			isNonProfitOrg = true;
```
is this `isNonProfitOrg` doing anything or can it be replaced by the above call to use hasNPCObjects()?

gabrielcouto357 replied:
isNonProfitOrg is used as a reference in the frSetup page to check which mapping options it should show.

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -41,8 +41,10 @@
     public static List<frMapping__c> mappings {
         get {
+            String objType = TYPE;
+            if(frUtil.hasNPCobjects()) objType = 'Gift Transaction';
```
Since this is changing the TYPE variable that is used in a few different places, is it possible to refactor the way TYPE is defined or called so we don't have to do this type of checking each time we access TYPE?   Right now the frSetupController is using the TYPE value defined in this class, but it is being similarly overwritten with different logic there.  

Would it be possible to have it such that there is a single frDonation.TYPE value and that resolves to the correct value based on the environment detected, so both the class that sets up the mappings and the class that uses the mapping (here) are referencing the same value?  That way we can be certain the queries are using the same criteria and it is easier to refactor if it needs to be changed in the future

gabrielcouto357 replied:
Added this in the new update! Just checks if its an org and assigns the type variable accordingly.

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +55,126 @@ public class frDonation extends frModel implements frSyncable {
-                Contact supporterContact = contacts.get(0);
-                o.fr_Donor__c = supporterContact.Id;
-                if(frUtil.hasNPSP()){
-                    o.put('npsp__Primary_Contact__c', supporterContact.Id);
```
It looks like this code has been removed,  but it shouldn't have been.  We should still be supporting if the NPSP is installed and populating this lookup field

gabrielcouto357 replied:
Sorry missed this! Added back in new update, if im missing any other existing functionality lmk and ill get right on it ty!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +55,126 @@ public class frDonation extends frModel implements frSyncable {
+            List<SObject> subscriptions = Database.query('SELECT Id FROM ' + subsciptionObjName + ' WHERE fr_ID__c = :subscriptionId');
             if (subscriptions.size() > 0) {
-                o.Subscription__c = subscriptions.get(0).Id;
+                o.put('GiftCommitmentId', subscriptions.get(0).get('Id'));
```
It looks like this removed the the non-NPC logic where we should be populating the Subscription__c field on the opportunity when we aren't in an NPC environment

gabrielcouto357 replied:
Added in new update!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +55,126 @@ public class frDonation extends frModel implements frSyncable {
+
+        //upsert
         try {
-            Database.upsert(opp, Opportunity.Fields.fr_Id__c, true);
```
Removing this upsert on the Funraise ID means we will create duplicate opportunities when the same Funraise record is synced multiple times, because it doesn't look like there is a query to find an existing opp for the same fr_Id__c value.  Ideally we won't have to have a query to find the existing record and we can restore this upsert

gabrielcouto357 replied:
Added this back for upserting opportunities! we check for existing gift transactions but this can be changed if needed!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -194,7 +238,7 @@ public class frDonation extends frModel implements frSyncable {
    
     
-    private void createOpportunityMapping(Map<String, Object> request) {
+    public void createOpportunityMapping(Map<String, Object> request) {
```
is there a reason we need to change this method to public?

gabrielcouto357 replied:
For visibility purposes to test in an NPC org, without it would need to make more modifications to change this but can look into it if needed!

Alexalex1432 replied:
I'm not seeing any usages of this method outside of this class, can you point out where it is being used in the test class?  If not, can we change this back to private?

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +55,126 @@ public class frDonation extends frModel implements frSyncable {
+        }
+
+        if(Test.isRunningTest() && String.valueOf(request.get('APEX_TESTS')) == 'true' && frUtil.hasNPCobjects()) {
+            setOpportunityStage(new Opportunity(), 'Closed Won');
```
Can you explain why we are setting the stage on a new Opportunity here instead of the object we are working with `o` ?

gabrielcouto357 replied:
This is for testing purposes! since in an NPC org this obj would not be an opportunity but we still need to test it for coverage

Alexalex1432 replied:
The status values being passed in here are incorrect.  Currently these are passing Opportunity Stage names, but the method setOpportunityStage operates off Status values that are different.  Only Complete/Failed/Pending/Refunded would provide code coverage

Alexalex1432 commented on src/classes/frDonor.cls:
```diff
@@ -64,90 +67,117 @@ public class frDonor extends frModel implements frSyncable{
+        String emailField = isPersonAccount ? 'PersonEmail' : 'Email';
+    
+        // Match based on email and names
+        if (supporter == null && String.isNotBlank(email)) {
```
It looks like this changed the logic that was there previously.  Previously, the matching logic would:
Try to query on first name, last name, and email
If no match is found, then query just on email

This would allow the matching logic to get a more specific match (where names match) when there are multiple records with the same email.  This change has made it so that if the names and email don't match, then there won't be a match

gabrielcouto357 replied:
Added this back in the new update!

Alexalex1432 replied:
Hi team, this looks like it was re-added or regressed back.  this block should not be entered if the firstname and lastname is blank.   Here is the previous logic that should be restored:
![image](https://github.com/user-attachments/assets/60e86c24-54b8-4973-962b-263309a9e15e)

mloney-yc replied:
@Alexalex1432 This logic has been restored.

gabrielcouto357 committed:
Apex class changes from recent feedback/internal review

Alexalex1432 commented:
Hi team, it looks like there was an issue committing that led to some files getting duplicated outside of the regular src directory:
![image](https://github.com/user-attachments/assets/9f764d7e-bd25-4e82-92c6-62fc45ea1653)

gabrielcouto357 committed:
Changing file location

gabrielcouto357 committed:
Delete frDonation.cls

gabrielcouto357 committed:
Delete frDonationTest.cls

gabrielcouto357 committed:
Delete frDonor.cls

gabrielcouto357 committed:
Delete frDonorTest.cls

gabrielcouto357 committed:
Delete frModel.cls

gabrielcouto357 committed:
Delete frPledge.cls

gabrielcouto357 committed:
Delete frSetupController.cls

gabrielcouto357 committed:
Delete frSetupControllerTest.cls

gabrielcouto357 committed:
Delete frSubscription.cls

gabrielcouto357 committed:
Delete frSubscriptionTest.cls

gabrielcouto357 committed:
Delete frUtil.cls

gabrielcouto357 commented:
Hello! Sorry about that, haven't had that one happen before! Updates should be pushed to the right spot and deleted the misplaced files. TY!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -41,8 +41,9 @@
 public class frDonation extends frModel implements frSyncable {
     public static List<frMapping__c> mappings {
         get {
+            String objType = TYPE;
```
It looks like this change is not doing anything now, can we remove this variable and restore the query to how it was previously?

gabrielcouto357 replied:
TYPE is now dynamically being set to the correct obj!

Alexalex1432 replied:
Sorry i meant the `objType` variable looks like it is unnecessary, as we're just re-assigning a variable

Alexalex1432 replied:
Hi team, can this variable be removed?

gabrielcouto357 replied:
Variable has been be removed in a previous update!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,136 @@ public class frDonation extends frModel implements frSyncable {
         Boolean result = false;
         Map<String, Object> request = getRequestBody();
-        o = new Opportunity();
+        for (String key : request.keySet()) {
```
Can this debug loop be removed?

gabrielcouto357 replied:
Removed!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,136 @@ public class frDonation extends frModel implements frSyncable {
+            o.put('AcknowledgementDate', System.today());
+        }
+
         applyMappings(o, request);
```
it looks like this method ends up getting called twice for non-NPC orgs, once on line 74 and once here.  Can we refactor this so that in any code path we're only applying the mappings once?

gabrielcouto357 replied:
Refactored to run once!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,136 @@ public class frDonation extends frModel implements frSyncable {
+            else {
+                o = gtList.get(0);
+            }
+            o.put('Name', String.valueOf(request.get('name')));
```
Do we need to enforce these in all cases or can they instead be replaced by a set of default mappings for the NPC orgs the same way we have some default mappings currently?  That way these can be overridden if a custom wants a different value for the name, payment method, etc

gabrielcouto357 replied:
Is there any preferences for the mapping defaults for the new objects? If not i can try and mimic the mapping to their non NPC counterparts.

gabrielcouto357 replied:
Thought i would ping this again, currently just have the default mapping set to creation time for acknowledgment date.

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,136 @@ public class frDonation extends frModel implements frSyncable {
+            o.put('Name', String.valueOf(request.get('name')));
+            o.put('OriginalAmount', (Decimal) request.get('amount'));
+            o.put('PaymentMethod', String.valueOf(request.get('paymentMethodType')));
+            o.put('AcknowledgementDate', System.today());
```
I think using the System.today() value here is incorrect as Funraise user has the ability to resync a transaction, or if an update is made to the transaction it will sync to Salesforce which will hit this code path.  That will end up changing the AcknowledgementDate to whenever the most recent sync was, rather than the original date of the gift

gabrielcouto357 replied:
AcknowledgementDate is a required field for gift transaction so we need to ensure this date is populated with something, is there a donation request field you could suggest to populate this?

Alexalex1432 replied:
Ok got it, so the way we have handled that for opportunity and contact is that the default mappings we provide handle setting all of the required fields (Stage, CloseDate, Name on Opportunity).  So I think this should be replaced with suggested default mappings in frSetupController

gabrielcouto357 replied:
Removed this and added defaults for the objects!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,136 @@ public class frDonation extends frModel implements frSyncable {
+            else supporters = [SELECT Id, AccountId FROM Contact WHERE fr_ID__c = :supporterFunraiseId];
+            if (supporters.size() > 0) {
+                SObject supporter = supporters.get(0);
+                o.put('DonorId', supporter.get('Id'));
```
I don't think this field will exist in the non-NPC case, as there isn't a standard field with this name on the Opportunity object.  The removed line 75 in this case is a key piece of functionality where the Opportunity and Contact are linked

gabrielcouto357 replied:
Changed to use the correct field in non NPC orgs!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,136 @@ public class frDonation extends frModel implements frSyncable {
+                    String pledgeId = npcSetting.NPCFieldName__c;
+                    o.put('Funraise_GT_Pledge__c', pledge.id);
+                } else {
+                    System.debug('NPC Setting not found at the organization level.');
```
I think we should remove this System.debug and replace it with some logging using the funraise_Error__c object if there is something we need the user to do

gabrielcouto357 replied:
Changed to use the frUtil.logRelationshipError method!

Alexalex1432 commented on src/classes/frDonor.cls:
```diff
@@ -64,90 +66,134 @@ public class frDonor extends frModel implements frSyncable{
+        String frId = getFunraiseId();
+    
+        SObject supporter = null; // Generic SObject for both Account and Contact
+        String sObjectType; // To track whether we're dealing with Account or Contact
```
Can this be removed and any usage replaced with final TYPE variable we have defined in this class?  
EDIT: I think that would also clean up some more code in the frSetupController class so that the TYPE values we have defined in frDonor and frDonation correspond with the SObject names and should be used in any places where we are referencing the SObjects for a donor and donation respectively

gabrielcouto357 replied:
Type would be Donor here so i used the sObjectType to limit calls to frUtil and double as a easy way to call Contact vs account.

mloney-yc replied:
@Alexalex1432 The TYPE in frDonor and frDonation aren't 1:1 with sObject types since they're different in NPC vs non-NPC. Are you recommending that we store an additional sobject type variable on frDonor and frDonation?

Alexalex1432 replied:
Yes I think that would be a great approach.  I think we're repeating logic in many of the classes to check what the correct SObject is to query, so I think it would be much better to centralize that logic and have a single constant reference of what SObject is used for these representations

Alexalex1432 commented on src/classes/frDonor.cls:
```diff
@@ -64,90 +66,134 @@ public class frDonor extends frModel implements frSyncable{
+        SObject supporter = null; // Generic SObject for both Account and Contact
+        String sObjectType; // To track whether we're dealing with Account or Contact
         
+        Boolean isPersonAccount = frUtil.hasNPCobjects();
```
for a bit more clarity on my end, can we rename this boolean to isNPC or something similar?  Just to make it clear that we are branching logic based on NPC/non-NPC environment

gabrielcouto357 replied:
Changed to isNPC!

Alexalex1432 commented on src/classes/frDonor.cls:
```diff
@@ -64,90 +66,134 @@ public class frDonor extends frModel implements frSyncable{
+            supporter.put('LastName', String.valueOf(request.get('institutionName')));
         }
-        
+        Database.upsert(supporter, true);
```
Can you review this saving behavior?  It looks like we upsert once here, and then have some branching logic on lines 185-188.  Additionally, it looks like the upsert that uses the fr_ID__c external ID field was removed, but I don't think it should have been removed as that is needed to not duplicate records

gabrielcouto357 replied:
Reverted behavior and uses external id field!

Alexalex1432 commented on src/classes/frDonor.cls:
```diff
@@ -159,10 +205,23 @@ public class frDonor extends frModel implements frSyncable{
+    private Account applyAccountMatch(List<Account> matchResults, String frId) {
+        if(matchResults != null && matchResults.size() > 0) {
+            Account donor = matchResults.get(0);
+            donor.fr_ID__c = frId;
```
It looks like this is a new custom field on Account but does not exist in this repo currently, so it wouldn't deploy

gabrielcouto357 replied:
Added field to Account Obj!

Alexalex1432 commented on src/classes/frModel.cls:
```diff
@@ -10,7 +10,7 @@ public with sharing abstract class frModel {
     public frModel(Sync_Attempt__c attempt) {
         this.syncRecord = attempt;
-        createLogRecord = attempt.Attempts__c >= frSyncRequestHandler.MAX_ATTEMPTS;
+        createLogRecord = attempt.Attempts__c >= 3;
```
Can we revert this change so the MAX_ATTEMPTS variable is used?

gabrielcouto357 replied:
Reverted!

Alexalex1432 commented on src/classes/frModel.cls:
```diff
@@ -120,6 +122,8 @@ public with sharing abstract class frModel {
             if(getSalesforceId() != null) {
                 query += 'Id = :salesforceId';
+            } else if(Test.isRunningTest()){
+                query += 'fr_Id__c = :funraiseId';
```
I don't think this is a valid change that should be included

gabrielcouto357 replied:
This is used for testing, i can attempt to look for a workaround but may take time!

Alexalex1432 replied:
If a test needs this to pass, then it sounds like there is a logic error elsewhere and we shouldn't patch just the test, but the problem in logic that is happening.

gabrielcouto357 replied:
Removed!

Alexalex1432 commented on src/classes/frPledge.cls:
```diff
@@ -49,6 +49,18 @@ public with sharing class frPledge {
+
+    public static Pledge__c findActiveAcc(Id accountId, Date closeDate) {
+        List<Pledge__c> possiblePledges = [SELECT Id FROM Pledge__c 
+                                         WHERE Supporter_PA__c = :accountId 
```
The lookup field referenced here has not been added to the repository so this would have errors deploying

gabrielcouto357 replied:
Added Account lookup for Pledge!

gabrielcouto357 committed:
Add files via upload

gabrielcouto357 committed:
Added Account lookup for Pledge

gabrielcouto357 committed:
Updates based on feedback

gabrielcouto357 committed:
Feedback updates

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,133 @@ public class frDonation extends frModel implements frSyncable {
-            Pledge__c pledge = frPledge.findActive(opp.fr_Donor__c, opp.CloseDate);
-            opp.Funraise_Pledge__c = pledge != null ? pledge.Id : null;
+        if(frUtil.hasNPCobjects()) {
+            List<SObject> alreadyMappedToPledgeList= Database.query('SELECT id FROM GiftTransaction WHERE fr_Id__c = :funraiseId AND Funraise_Pledge__c != null');
```
It seems like this query is re-querying the record we already have in memory (based on the fr_Id__c value).  Can we update the query on line 74 to add in selecting the column Funraise_Pledge__c and that way we don't need to perform another query here, we can just check if Funraise_Pledge__c is populated on the `o` object

gabrielcouto357 replied:
This has been removed!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -338,7 +407,7 @@ public class frDonation extends frModel implements frSyncable {
     ///CONSTANTS///
 
-    public static final String TYPE = 'Donation';
+    public static final String TYPE = frUtil.hasNPCobjects() ?  'Gift Transaction' : 'Donation';
```
Is there a need for this value to be different based on if the NPC objects are present or not?  In my understanding there isn't a way to convert an existing Salesforce org between NPC vs non-NPC, so the mappings belong to donations, not Opportunity/GiftTransaction.  I think removing this can simplify a lot of the mapping and frSetupController code

EDIT: this also applies to the TYPE in frDonor

gabrielcouto357 replied:
This could be fixed but some of the tests/ functionality is tied to this so may take some time to revert it.

Alexalex1432 replied:
I think both of the TYPE variables should not be changed for this pull request, since these are not used in queries and previously were not `Opportunity` nor `Contact` I don't think they need to correspond with the SObject

gabrielcouto357 replied:
Removed in a previous update!

Alexalex1432 commented on src/classes/frDonorTest.cls:
```diff
@@ -56,8 +56,13 @@ public class frDonorTest {
+            System.assertEquals(1, newContact, 'The funraise donor id was not populated to the contact field');
+        }
+        else {
+            Integer newContact = [SELECT COUNT() FROM Account];
```
If there isn't anything being asserted from this, do we nee dthis change?  It seems like this test should be asserting that the donor was created as an Account/PersonAccount 

EDIT: There appears to be a pattern in this class that we are not asserting results in the NPC environment.  Can we update the tests so that we are asserting records are created as expected?  Similar to the asserts in the existing/non-NPC case

gabrielcouto357 replied:
This ill work on updating, but shouldn't affect the tests or anything timeline wise. LMK if this should be prioritized!

gabrielcouto357 replied:
Updated along with the other tests!

Alexalex1432 commented on src/classes/frModel.cls:
```diff
@@ -78,14 +78,16 @@ public with sharing abstract class frModel {
         Boolean needLastModifiedDate = false;
-        
-        for(frMapping__c mapping : getMappings()) {
+        List<frMapping__c> objMappings = getMappings();
```
I don't follow why we call getMappings() here but then re-query them.  Shouldn't getMappings() handle getting the specific type that we need so that we don't need to add branching logic in this base class for each possible type?

gabrielcouto357 replied:
Fixed! Except gift commitment wouldnt work without it, so leaving that there for now. No functionality should be affected while its there for now while looking for a solution

Alexalex1432 replied:
Ok, got it.  I do think that should be resolvable so we'll leave this open until this re-querying can be removed

gabrielcouto357 replied:
I could add getMappings, mappings and a TYPE to frSubscription and get mappings would work here. But this would lead to an unnecessary soql call every time a subscription syncs over in a nonNPC enviro so thinking overall the one line soql call would be better unless theres plans of introducing mapping to nonNPC for subscriptions

mloney-yc replied:
@Alexalex1432 getMappings() has been added to frSubscription and this loop has been reverted back to loop over getMappings().

Alexalex1432 commented on src/classes/frModel.cls:
```diff
@@ -214,10 +217,7 @@ public with sharing abstract class frModel {
-                                +' Record Id: '+record.Id+' - Funraise Id: '+ funraiseId + ' - Field: '+field.getDescribe().getName()+' - Value: '+value
-                                +' Exception: '+ex
-                               );
+            System.debug(ex);
```
This change should be reversed, the error should be logged in the custom object and not debugged

gabrielcouto357 replied:
Added back this with an exception to just debug it as i was running into some null pointer errors in some tests files.

Alexalex1432 commented on src/classes/frSetupController.cls:
```diff
@@ -69,6 +87,19 @@ public with sharing class frSetupController {
 	public List<SelectOption> donorFROptions {get; set;}
 	public List<frMapping__c> donorMappings {get; set;}
+
+	public List<SelectOption> accSFOptions {get; set;}
```
I don't think we should be duplicating the selectoption and frMapping variables here.  The above mapping variables aren't specific to Opportunity/Contact, we should just change what options are in there.

gabrielcouto357 replied:
This was to make the code a bit easier to read/debug. This can be removed if needed but wont affect functionality that im aware of.

Alexalex1432 replied:
yes, please remove.  

I think having added variables that are duplicating things the other variables are doing makes it a bit more confusing and harder to manage (especially because there are not comments that explain why there are so many options that would not be used in certain environments) rather than having each Funraise object having a single select list of options, and the object used to populate those objects is what changes per environment (NPSP vs NPC vs neither).   

The gift commitment options are the only ones that I think would need to be added since there isn't currently mappable options for subscriptions/commitments

gabrielcouto357 replied:
Updated!

Alexalex1432 commented on src/classes/frSubscription.cls:
```diff
@@ -71,24 +71,74 @@ public class frSubscription extends frModel implements frSyncable {
     private Boolean create(Map<String, Object> request) {
         Boolean result = false;
         String funraiseId = getFunraiseId();
+        for (String key : request.keySet()) {
```
debug needs to be removed here

gabrielcouto357 replied:
Removed!

gabrielcouto357 committed:
Delete src/objects/NPC_Setting__c.object

gabrielcouto357 committed:
Update package.xml to add <members>Account.fr_ID__c</members>

gabrielcouto357 committed:
Added Commitment  type mappings for commitment mapping in frSetup

gabrielcouto357 committed:
Changed to include correct mapping for commitments

gabrielcouto357 committed:
Updated to avoid some dependency issues when not in NPC Org

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,139 @@ public class frDonation extends frModel implements frSyncable {
         String funraiseId = getFunraiseId();       
-        String supporterFunraiseId = String.valueOf(request.get('donorId'));        
+        String supporterFunraiseId = String.valueOf(request.get('donorId')); 
+        String supporterObjName = frUtil.hasNPCobjects() ? 'Account' : 'Contact';
```
This should be pulled from frDonor.TYPE

gabrielcouto357 replied:
I could use the .Type in replacement for the frUtil.hasNPCobjects(), but since its not the object API name i still need to use this kind of statement

mloney-yc replied:
@Alexalex1432 Since frDonor.TYPE is not an sobject type and doesn't map 1:1 with an sobject type, would you prefer that we add a new variable to frDonor to track the sobject type?

Alexalex1432 replied:
Yes I think that would be a very clean approach and allow for easier refactoring or substitutions if more sobjects would ever need to be implemented for these (frDonation, frDonor, frSubscription etc)

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,139 @@ public class frDonation extends frModel implements frSyncable {
-        String supporterFunraiseId = String.valueOf(request.get('donorId'));        
+        String supporterFunraiseId = String.valueOf(request.get('donorId')); 
+        String supporterObjName = frUtil.hasNPCobjects() ? 'Account' : 'Contact';
+        String subsciptionObjName = frUtil.hasNPCobjects() ? 'GiftCommitment' : 'Subscription__c';
```
This should be pulled from frSubscription.TYPE

gabrielcouto357 replied:
I could use the .Type in replacement for the frUtil.hasNPCobjects(), but since its not the object API name i still need to use this kind of statement.

mloney-yc replied:
@Alexalex1432 Since frSubscription.TYPE is not an sobject type and doesn't map 1:1 with an sobject type, would you prefer that we add a new variable to frSubscription to track the sobject type?

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +54,139 @@ public class frDonation extends frModel implements frSyncable {
-            frPledge.create(opp);
+
+        if(isPledge && o.get('Id') != null) {
+            System.debug('In create pledge');
```
This debug needs to be removed

gabrielcouto357 replied:
Removed!

gabrielcouto357 committed:
Add files via upload

Alexalex1432 commented on src/customMetadata/frField.supported.md:
Can you explain what this frField is for?  There isn't something synced from Funraise called `supported`

gabrielcouto357 replied:
Just random fields i found on Subscription upload, we added mapping to it. If you had a preferred list for this object i can add that as well!

Alexalex1432 commented on src/customMetadata/frField.id.md:
As previously discussed, the Funraise ID field should not be a field that users should map, this should always be handled by the integration populating the fr_Id__c field.  This file should be removed

gabrielcouto357 replied:
Removed, and how should we handle this instead? How would you like to handle this instead, maybe field validation checking to ensure this field exists before allowing mappings to be saved?

Alexalex1432 commented on src/customMetadata/frField.deleted.md:
This should not be a mappable field and instead handled by the syncing classes, such as in frSubscription where the record is deleted when it is set to true.

gabrielcouto357 replied:
Deleted this field!

Alexalex1432 commented on src/classes/frSubscription.cls:
```diff
@@ -74,21 +74,83 @@ public class frSubscription extends frModel implements frSyncable {
+                }
+            }
+
+            String strPledgeAmount2 = String.valueOf(request.get('pledgeAmount'));
```
Can we fix the variable naming here, I'm not following why this is called strPledgeAmount2

gabrielcouto357 replied:
Fixed!

Alexalex1432 commented on src/classes/frSubscription.cls:
This class is not handling the delete properly in an NPC environment

gabrielcouto357 replied:
Fixed!

gabrielcouto357 committed:
Apex updates from feedback

gabrielcouto357 committed:
Delete src/customMetadata/frField.deleted.md

gabrielcouto357 committed:
Delete src/customMetadata/frField.id.md

Alexalex1432 commented on src/classes/frDonor.cls:
```diff
@@ -39,7 +39,7 @@
 
 public class frDonor extends frModel implements frSyncable{
-    public static final String TYPE = 'Donor';
+    public static final String TYPE = frUtil.hasNPCobjects() ? 'Person Account' : 'Donor';
```
Note that previously this was not `Contact`, so there shouldn't be a reason to change the Type away from `Donor`, as this is still a Donor type, just using a different underlying SObject

gabrielcouto357 replied:
Removed!

Alexalex1432 commented on src/classes/frDonor.cls:
```diff
@@ -56,6 +56,7 @@ public class frDonor extends frModel implements frSyncable{
     }
     
     private Contact supporterContact;
+    private Account supporterAccount;
```
Can you explain the need for this variable?  It does not appear to be populated currently which causes other issues in the place where the ID is being read

gabrielcouto357 replied:
Added the assignment for these 2 objects to allow the id to be read!

Alexalex1432 replied:
I'm not following the logic for this object to be introduced, it looks like all of the operations in this class are using the generic sObject `supporter`.  the supporterAccount and even this existing supporterContact object don't seem to be used aside from being assigned to that sObject.  Is there any reason we shouldn't just remove both of these objects to be replaced by a single sObject?

mloney-yc replied:
@Alexalex1432 This has been updated to use a generic sobject for supporter

Alexalex1432 commented on src/classes/frModel.cls:
```diff
@@ -214,10 +213,14 @@ public with sharing abstract class frModel {
                                 +' Record Id: '+record.Id+' - Funraise Id: '+ funraiseId + ' - Field: '+field.getDescribe().getName()+' - Value: '+value
                                 +' Exception: '+ex
                                );
+            } catch (Exception ignore) {
```
Was there an exception being encountered on inserting the Error__c record?  I think that's concerning and should be addressed rather than catching an exception and ignoring it here, or just debugging it

I do think this should be removed or if there is an exception being encountered on inserting the error we should address that, as this change as it is now does not work (and the exception is being ignored)

gabrielcouto357 replied:
This was only occurring in test cases as it was making mappings of 'Donation' Type but for fields present on an opp only. On FP when the tests are redone i will be able to remove this!

gabrielcouto357 replied:
Updated!

Alexalex1432 commented on src/classes/frSetupController.cls:
```diff
@@ -5,6 +5,24 @@ public with sharing class frSetupController {
 		}
 		private set;
 	}
+	public static String PERSON_ACCOUNT_TYPE {
```
This should be removed and the shared variable that defines the type in frDonation, frDonor etc should be used.

gabrielcouto357 replied:
Removed!

Alexalex1432 commented on src/classes/frSetupController.cls:
```diff
@@ -90,7 +171,7 @@ public with sharing class frSetupController {
 		}
 		donationSFOptions.sort();
-		for(frField__mdt field : [SELECT DeveloperName, MasterLabel FROM frField__mdt WHERE Type__c = :DONATION_TYPE ORDER BY MasterLabel]) {
+		for(frField__mdt field : [SELECT DeveloperName, MasterLabel FROM frField__mdt WHERE Type__c = 'Donation' OR Type__c = 'All' ORDER BY MasterLabel]) {
```
The ALL type mapping needs to be removed, there isn't a concept of frFields that apply to all objects currently

gabrielcouto357 replied:
Removed from all queries and as an option on the object

Alexalex1432 commented on src/classes/frSetupController.cls:
```diff
@@ -157,21 +260,23 @@ public with sharing class frSetupController {
 					mapping.Name = mapping.Name.subString(0, customSettingNameLength);
 				}
-				upsertList.add(mapping);
+				if(!upsertList.contains(mapping)) upsertList.add(mapping);
```
Can you explain why this was added?  It seems like this is trying to prevent a possible duplicated mapping, but I don't think this is the right way to handle that edge case as it will just silently remove a mapping that the user thinks will be saved.  I think this should be removed or if there is an error that is requiring it, can you share that here?

gabrielcouto357 replied:
Removed for now, but did notice that an error could occur here if a user tried to insert a mapping record with a duplicate sf field.

Alexalex1432 commented on src/classes/frSetupControllerTest.cls:
```diff
@@ -5,55 +5,94 @@ public class frSetupControllerTest {
 		Set<String> expectedDonationFieldNames = new Set<String>();
 		for(frField__mdt field : [SELECT DeveloperName, Type__c FROM frField__mdt]) {
-			if(frDonor.TYPE.equals(field.Type__c)) {
+			if('Donor'.equals(field.Type__c)) {
```
These changes are not correct and the frDonor/frDonation TYPE variables should be used here.  That was the purpose of having that publicly defined variable that can be shared and managed in one place, so we don't have to update strings in various places

gabrielcouto357 replied:
Changed!

Alexalex1432 commented on src/objects/Pledge__c.object:
```diff
@@ -197,6 +196,21 @@
+        <relationshipOrder>0</relationshipOrder>
+        <reparentableMasterDetail>false</reparentableMasterDetail>
+        <trackTrending>false</trackTrending>
+        <type>MasterDetail</type>
```
BETA PACKAGE BLOCKER :warning: : 

We cannot add a MasterDetail field to an object that won't be populated in certain environments.  Master Detail fields are required and NPSP or non-person account environments will not be populating this value.  

This change is also responsible for some of the current deploy issues.

gabrielcouto357 replied:
Changed to lookup!

Alexalex1432 commented on src/pages/frSetup.page:
```diff
@@ -131,5 +204,119 @@
                 </apex:commandLink>
             </apex:pageBlockSection>
         </apex:pageBlock>
+        <apex:pageBlock title="Donor --> Person Account" rendered="{!personAccountsEnabled == true}">
```
Do we actually need to have a separate table for these, or can it be the case that once we remove the duplicated option variables from the controller, the existing table can be used but just have the instructions on the page change to reflect the correct objects (like showing Person Account instead of Contact in this pageBlock title)

gabrielcouto357 replied:
I would think this would be better separately as combining them would mean including custom labels or adding more methods in the controller to control what text to show making it more complex overall.

gabrielcouto357 committed:
Updates based on feedback

gabrielcouto357 committed:
Update objects for correct field type and picklist option

gabrielcouto357 committed:
Updated page to use correct types

gabrielcouto357 committed:
Updated to include frTestUtil and added the Sync object

Alexalex1432 commented on src/package.xml:
```diff
@@ -86,6 +88,7 @@
         <members>Subscription__c</members>
         <members>frField__mdt</members>
         <members>frMapping__c</members>
+        <members>Sync_Attempt__c</members>
```
It looks like this object might be missing from the repo, and adding it here started to cause a deployment failure:
BUILD FAILED
*********** DEPLOYMENT FAILED ***********
Request ID: 0Af6R00000coIAUSA2

All Component Failures:
1.  package.xml (Sync_Attempt__c) -- Error: An object 'Sync_Attempt__c' of type CustomObject was named in package.xml, but was not found in zipped directory

*********** DEPLOYMENT FAILED ***********

gabrielcouto357 committed:
Added Sync Object

Alexalex1432 commented:
leaving a note here that there are failing tests that prevent the creation of the beta package

gabrielcouto357 committed:
Test Class fixes

gabrielcouto357 committed:
Test Class fixes

gabrielcouto357 committed:
Fix frSubscription Test

gabrielcouto357 committed:
Update frSubscriptionTest

gabrielcouto357 committed:
Uncommented test class

gabrielcouto357 committed:
Updated to correct type__c for frMapping SOQL query

gabrielcouto357 committed:
Add frAnswer code to allow lookup to Person Accounts

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +53,137 @@ public class frDonation extends frModel implements frSyncable {
+            setOpportunityStage((Opportunity) o, String.valueOf(request.get('status')));
+        } 
+        else {
+            List<SObject> gtList = Database.query('SELECT Id, Name, DonorId, AcknowledgementDate, CheckDate FROM GiftTransaction WHERE fr_ID__c = :frId');
```
Do we need to query for the existing object here rather than just instantiating the sobject, similar to how it works for Opportunity?  If we do have to do this query, do we need these fields that are being selected for any reason?

gabrielcouto357 replied:
Doesn't seem to work without it and were not close to any soql limits so thinking this can stay, removed the extra fields!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +53,137 @@ public class frDonation extends frModel implements frSyncable {
+            List<SObject> subscriptions = Database.query('SELECT Id FROM ' + subsciptionObjName + ' WHERE fr_ID__c = :subscriptionId');
             if (subscriptions.size() > 0) {
-                o.Subscription__c = subscriptions.get(0).Id;
+                if(!frUtil.hasNPCobjects()) o.put('Subscription__c', subscriptions.get(0).get('Id'));
```
Since we're already doing the statement above to figure out the object name, should we follow the same pattern here to set the field name?  It seems like it would clean up this code a bit and reduce the lines for test coverage

gabrielcouto357 replied:
Not sure what you mean here, would you mean using the fieldName instead of !frUtil.hasNPCobjects()?

mloney-yc replied:
@Alexalex1432 This has been moved into a variable higher up like `supporterObjName` and `subsciptionObjName`.

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +53,137 @@ public class frDonation extends frModel implements frSyncable {
+            if(!isPledge && !alreadyMappedToPledge) {
+                Pledge__c pledge = frPledge.findActiveAcc((Id) o.get('DonorId'), (Date) o.get('CheckDate'));
+                if (pledge != null) {
+                    try {
```
Is there a reason for this to be wrapped in a try-catch?  It also seems like the exception being logged in the catch block is incorrect and logging issues around the Subscription ID when trying to map an opportunity to a pledge

gabrielcouto357 replied:
This has been removed!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +53,137 @@ public class frDonation extends frModel implements frSyncable {
+                        frUtil.logException(getFrType(), funraiseId, ex);
+                    }
+                } else {
+                    frUtil.logRelationshipError(getFrType(), funraiseId, 
```
Why is an error being logged here if there is no active pledge?

gabrielcouto357 replied:
This has been removed!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +53,137 @@ public class frDonation extends frModel implements frSyncable {
+            createOpportunityMapping(request);
+        }
+
+        if(Test.isRunningTest() && String.valueOf(request.get('APEX_TESTS')) == 'true' && frUtil.hasNPCobjects()) {
```
I don't think this is a valid way to get test coverage by introducing production code that is only run in tests.  Can you remove this and investigate a different way that does not compromise code to achieve test coverage?

gabrielcouto357 replied:
Needed more test coverage in non-NPC environments so needed to touch some of the NPC code to hit the 75%, Dont believe theres another way since i cant change org settings within a test, ie activated PA and adding NPC objects

Alexalex1432 replied:
Can we use https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation_testvisible.htm the testVisible annotation on the method being covered here to be able to call it directly in the test class rather than having this in the production logic?

mloney-yc replied:
@Alexalex1432 This has been removed and code coverage is achieved with the testVisible annotation and a new test method.

Alexalex1432 commented on src/classes/frDonationTest.cls:
As previously shared, the tests in this class have been modified but do not assert on any results.  I don't think the existing methods should be changed to support tests in both environments, but rather there should be a set of tests that are for Person-Account/NPC environments, and tests for other environments.  

Adding branches in the test setup and (missing) assertions make the test methods harder to read and does not clearly indicate failures if the tests stop passing

gabrielcouto357 replied:
Added the assertions to match for both NPC and non NPC orgs!

gabrielcouto357 committed:
Updated tests along with feedback updates for frDonation

gabrielcouto357 committed:
Add files via upload

gabrielcouto357 committed:
Updates to frDonation

gabrielcouto357 committed:
Various updates

Redid tests to separate them from NPC to non NPC, updates for model to ensure errors get logged and updates on the frSetup Controller

gabrielcouto357 committed:
Updates to frSetup to reduce variable count

gabrielcouto357 commented:
> As a general note, can we remove any System.debug or commented out code that won't be needed. If there are any comments that you need input from me on, please tag in a comment here

Sorry didn't respond to this one, but this has been updated in a previous update!

gabrielcouto357 commented:
> We will need to update the `getFields()` and `getObjects()` method in any class that extends frModel that were changed here, as these are used for checking permissions and that was a required change from Salesforce as part of our App Security review, that we need to ensure we have permissions for the objects/fields that will be modified.

> in frUtil we have a method called `hasNPSP()` to check if the organization has the nonprofit success pack. Can we refactor the current schema checks for NPC functionality to a similar centralized method so it can be more easily managed and clearly show we are entering a blocks of logic for NPC

> `getSalesforceId()` should be updated to return the gift transaction ID in the NPC situations, as this will be used in various places, including field mappings that are queried for in `frModel.applyMappings` which are used to pull the current record and resolve data conflicts between Funraise & Salesforce

> Discussed elsewhere, but I'll also leave a comment here to keep track that we need to add test coverage for this new logic and ensuring a sync from Funraise results in the correct records and data being created (along with providing the required coverage enforced by Salesforce)

Couldn't respond to these directly, but these have all been updated either now or in a previous update!

Alexalex1432 commented on src/classes/frDonation.cls:
```diff
@@ -53,84 +53,127 @@ public class frDonation extends frModel implements frSyncable {
+
+        if(!frUtil.hasNPCobjects()) {
+            o = new Opportunity();
+            setOpportunityStage((Opportunity) o, String.valueOf(request.get('status')));
```
Looking at the code diff here, we have changed the order of this logic.  Previously the order was 

(1) applyMappings
(2) setOpportunityStage

The setOpportunityStage method has logic to not set the stage if it was already set by the mappings, but now that logic would never execute since that is happening first.  I think we should restore the order of these methods

mloney-yc replied:
@Alexalex1432 The previous order has been restored.

Alexalex1432 commented on src/classes/frDonationTest.cls:
```diff
@@ -102,44 +187,98 @@ public class frDonationTest {
+            return;
+        }
+        // Original code for the NPC else-block didn't do anything, so we skip or do minimal logic
+        // If you want to add an actual test for status logic in NPC, you can do so here.
```
Is there an intent to add test code to this method? If not can it be removed?

mloney-yc replied:
@Alexalex1432 This test has been removed.

Alexalex1432 commented on src/classes/frDonationTest.cls:
```diff
@@ -161,30 +300,37 @@ public class frDonationTest {
+        ];
         List<OpportunityContactRole> roles = newOpportunity.OpportunityContactRoles;
-        System.assertEquals(3, roles.size(), 'Expected to be a role for donor, fundraiser, and team captain');
-        for(OpportunityContactRole role : roles) {
```
Were these lines removed for any reason?  It seems like good testing logic to have confirmed that the correct opportunity contact roles are being created for separate contacts

mloney-yc replied:
@Alexalex1432 All non-NPC tests have been restored, so this has been added back in.

Alexalex1432 commented on src/classes/frDonorTest.cls:
```diff
@@ -57,12 +63,42 @@ public class frDonorTest {
         
         String frId = String.valueOf(getTestRequest().get('id'));
         Integer newContact = [SELECT COUNT() FROM Contact WHERE fr_Id__c = :frId];
-        System.assertEquals(1, newContact, 'The funraise donor id was not populated to the contact field');
```
Why was this assert message removed?

mloney-yc replied:
@Alexalex1432 This has been restored.

Alexalex1432 commented on src/classes/frDonationTest.cls:
```diff
@@ -195,75 +341,183 @@ public class frDonationTest {
+
+        String oppFrId = String.valueOf(request.get('id'));
+        frTestUtil.assertNoErrors();
+        List<SObject> gt = Database.query('SELECT Id, Name, DonorId, GiftCommitmentId FROM GiftTransaction WHERE fr_ID__c = :oppFrId');
```
It looks like this is being queried for but it is not being used, is there a missing assert here?

mloney-yc replied:
@Alexalex1432 An assert has been added at the end of this test that uses this GiftTransaction.

Alexalex1432 replied:
Hi @mloney-yc 

It looks like this method is testing for a Subscription being connected (based on the method name) and I see a Subscription object is being created in the test.  Should the assert be confirming that the resulting gift transaction is connected to the Subscription lookup field?

gabrielcouto357 replied:
This is matching the existing test using this line System.assertEquals(gc[0].Id, gt[0].get('GiftCommitmentId')); where gc would be the subscription and GiftCommitmentId would be the field its stored on giftTransactions in a NPC environment

Alexalex1432 commented on src/classes/frPledge.cls:
```diff
@@ -86,4 +98,48 @@ public with sharing class frPledge {
+        Id donorId = (Id) gt.get('DonorId');
+        SObject acc = Database.query('SELECT PersonContactId FROM Account WHERE Id = :donorId');
+        Pledge__c pledge = new Pledge__c(
+            Name = frUtil.truncateToFieldLength(Pledge__c.Name.getDescribe(), (String) gt.get('Name') + ' Pledge'),
```
Is this name field on the gift transaction always going to be populated?  Should there be any fallback in case it isn't present?  Also it looks like it would be possible to throw a "SObject row retrieved via SOQL without querying the requested field" exception here if the field wasn't set via mappings or retrieved via the SOQL query, right?

mloney-yc replied:
@Alexalex1432 Name will always be set on the record. Additional checks have been added to handle `Name` and/or `OriginalAmount` not being set on a record due to not being included in a query (or any other reason).

mloney-yc committed:
Add dev files to .gitignore

mloney-yc committed:
Use generic sobject for supporter

mloney-yc committed:
Store subscription lookup field name in a var, like object name

mloney-yc committed:
Dedicated test for frDonation.setOpportunityStage()

mloney-yc committed:
Call setOpportunityStage after applyMappings

mloney-yc committed:
Revert non-NPC tests (frDonationTest, frDonorTest)

gabrielcouto357 committed:
Revert non-NPC tests frSubscription

gabrielcouto357 committed:
Frdonor change for email matching

mloney-yc committed:
Add getMappings() to frSubscription

mloney-yc committed:
Merge branch 'YC-NPC-Updates-fp' of github.com:funraise/Funraise-Salesforce into YC-NPC-Updates-fp

gabrielcouto357 committed:
Updated null handling in frPledge

mloney-yc committed:
Merge branch 'YC-NPC-Updates' of github.com:funraise/Funraise-Salesforce into YC-NPC-Updates

mloney-yc committed:
Separate NPC and non-NPC tests in frSetupControllerTest

mloney-yc committed:
Explain frSubscription.fields

mloney-yc committed:
Handle missing GiftTransaction fields

mloney-yc committed:
NPC tests all passing

mloney-yc committed:
Remove unused test

gabrielcouto357 committed:
Updates to fix deployment into non NPC org

Alexalex1432 commented on src/classes/frAnswer.cls:
```diff
@@ -77,27 +77,32 @@ public class frAnswer extends frModel implements frSyncable {
+            List<SObject> supporters;
+            if(frUtil.hasNPCobjects()) supporters = Database.query('SELECT Id FROM Account WHERE fr_ID__c = :funraiseSupporterId');
+            else supporters = [SELECT Id, AccountId FROM Contact WHERE fr_ID__c = :funraiseSupporterId];
+            System.debug('supporters: ' + supporters);
```
Can we remove this debug statement?

gabrielcouto357 replied:
Removed!

Alexalex1432 replied:
Hey team it looks like this debug was not removed

gabrielcouto357 replied:
This still looks to be removed can you check again please?

Alexalex1432 commented on src/classes/frModel.cls:
```diff
@@ -175,7 +173,7 @@ public with sharing abstract class frModel {
                 }
             } else if (field.getDescribe().getType() == Schema.DisplayType.DateTime) {
-                record.put(field, DateTime.newInstance((Long)value));
+                if(value != NULL) record.put(field, DateTime.newInstance((Long)value)); //Added null handling
```
I think this introduction may interfere with other logic.  ON the field mapping screen, there is an option under Conflict Resolutions for each field so that the users can choose if/when data is overwritten.  However with this change the result would be that they may choose for Funraise Data to Overwrite Salesforce data, but if the value is null in Funraise it would not overwrite, possibly leading to misrepresented values somewhere along the line.  

Can we instead change this so only if the value is not null then we try to parse it into a new DateTime, and otherwise we just write `null` ?

gabrielcouto357 replied:
Added an else to write the null value as expected!

Alexalex1432 replied:
@gabrielcouto357 Can you clarify why this change is needed at all?  In an instance where this would throw an exception, that exception would be caught below and route to the below write method which should work with setting a datetime to null.   

I ask because the other branches in this method are not doing this type of logic so it makes the behavior a bit less clear that this branch has special handling for null

gabrielcouto357 replied:
We were having some cases where some test cases and some lives cases were causing this line to error in particular. After adding these lines the tests passed and with the new additions this shouldn't affect functionality

Alexalex1432 commented on src/classes/frSubscription.cls:
```diff
@@ -74,21 +80,88 @@ public class frSubscription extends frModel implements frSyncable {
+                Supporter__c = contacts.get(0).Id
+            );
+        }
+        if(Test.isRunningTest() && frUtil.hasNPCobjects()) {
```
This change seems suspect to me, can we review if it is actually necessary?  In general I would like to take all steps we can to avoid having Test.isRunningTest() in production-executed code, and this block does not appear to be doing much meaningful work so perhaps it can be removed.

gabrielcouto357 replied:
This is necessary to add code coverage to frSubscription, the code block replicates the creation of a Subscription but uses the personcontactId to populate the field, code coverage is below the 75% without this addition

Alexalex1432 replied:
I don't think this is valid way to achieve test coverage, and this is creating an object that is not actually used, and no meaningful code is being tested.

gabrielcouto357 replied:
Removed! Coverage seems to still be above the 75% but let me know if this becomes an issue!

Alexalex1432 commented on src/classes/frSubscription.cls:
```diff
@@ -116,9 +189,11 @@ public class frSubscription extends frModel implements frSyncable {
             String salesforceFieldName = funraiseFieldName.replaceAll(regex, replacement);
             salesforceFieldName =+ 'funraise__' + salesforceFieldName.capitalize() + '__c';
             Schema.SObjectField field = fields.get(salesforceFieldName);
+            if(field == NULL) continue; //added null handling
```
Can we discuss why this was added?  In general I think the field being null would be pretty exceptional since it should be saved based on us introspecting the fields availalbe in frSetupController.  If we're going to add handling on this, can an exception be logged in a Error__c object

gabrielcouto357 replied:
Added for tests in NPC environments, changed the method to the frUtil.hasNPCObjects() as this wont affect functionality in either org this way!

gabrielcouto357 committed:
Updates based on feedback

Alexalex1432 commented on src/classes/frAnswer.cls:
Hey team can you look at what may have caused the most recent commit to appear as removing nearly all of the contents and replacing them?  That does make reviewing the changes a bit harder since it isn't fully clear what has changed.  A common culprit I've seen is if you have certain whitespace settings that conflict with what is already present

Alexalex1432 replied:
Hey all, it still looks like the first half of this class was fully removed/changed.  I wouldn't expect any changes in the top copyright block at all.

gabrielcouto357 replied:
Not sure what caused this or how to fix it, code looks good and is passing tests so thinking this is a git issue potentially?

Alexalex1432 commented on src/classes/frSubscription.cls:
```diff
@@ -74,21 +80,88 @@ public class frSubscription extends frModel implements frSyncable {
+            if(pledgeAmount1 != null) {
+                frPledge.create(gc, pledgeAmount1);
+            }
+            if(!Test.isRunningTest()) {
```
Can this logic being checked?  This does not look correct to me and it isn't clear why we are suppressing the result in a situation where a test is running

gabrielcouto357 replied:
Removed! Was for test coverage for non-NPC code in a NPC environment, coverage seems to still be above the 75% but let me know if this becomes an issue!

Alexalex1432 commented on src/classes/frSubscription.cls:
```diff
@@ -74,21 +80,88 @@ public class frSubscription extends frModel implements frSyncable {
-            Name = frUtil.truncateToFieldLength(
-                Subscription__c.Name.getDescribe(), 
+        Subscription__c subscription;
+        if(contacts.isEmpty() && frUtil.hasNPCobjects()) {
```
Can you clarify why is this branch is checking if contacts list are empty?  This seems like a logic branch that is specific to an NPC environment so that contacts list would ALWAYS be empty, correct?

gabrielcouto357 replied:
Not needed, was most likely from older logic, and has been removed!

Alexalex1432 commented on src/classes/frSubscription.cls:
```diff
@@ -116,9 +189,11 @@ public class frSubscription extends frModel implements frSyncable {
             String salesforceFieldName = funraiseFieldName.replaceAll(regex, replacement);
             salesforceFieldName =+ 'funraise__' + salesforceFieldName.capitalize() + '__c';
             Schema.SObjectField field = fields.get(salesforceFieldName);
+            if(frUtil.hasNPCobjects()) continue; //added handling for NPC orgs since we need to hit this for code coverage
```
This should be removed and is not a valid change

gabrielcouto357 replied:
Removed!

Alexalex1432 commented on src/classes/frSubscription.cls:
```diff
@@ -136,46 +211,74 @@ public class frSubscription extends frModel implements frSyncable {
     
     protected override frUtil.Entity getFrType() {
-        return frUtil.Entity.SUBSCRIPTION;
+        if(frUtil.hasNPCobjects()) return frUtil.Entity.GIFTCOMMITMENT;
```
Can you explain the need for this change?  In either environment, this will still always represent a Subscription, correct?  The other classes do not have this differentiation and I'm not sure it is needed.  If possible, it would keep the code cleaner to revert this

gabrielcouto357 replied:
reverted!

Alexalex1432 commented on src/classes/frSetupControllerTest.cls:
```diff
@@ -174,6 +220,7 @@ public class frSetupControllerTest {
 		}
 		//remove empty string from controller values since we don't care about the --None-- option
 		actualValues.remove('');
+		actualValues.remove('id');
```
Can you share why this change was added?

gabrielcouto357 replied:
Was left accidentally when editing the test class, removed!

gabrielcouto357 committed:
Update to remove .remove('id')

gabrielcouto357 committed:
Updates for frSubscription

gabrielcouto357 committed:
Deduped some variables related to request

gabrielcouto357 committed:
Updates for duplicate request variables

gabrielcouto357 committed:
Added a missing line for pledgeAmount

Alexalex1432 commented on src/classes/frDonationTest.cls:
```diff
@@ -195,85 +345,181 @@ public class frDonationTest {
-    
-    static testMethod void syncEntity_LinkToSubscription() {  
+
+    static testMethod void syncEntity_newOpportunityRoleMappings_missingContacts_NPC() {
```
Since this method is not testing anything, I think it should be removed

gabrielcouto357 replied:
Removed!

Alexalex1432 commented on src/classes/frDonationTest.cls:
```diff
@@ -195,85 +345,181 @@ public class frDonationTest {
+
+        Map<String, Object> request = getTestRequest();
+        request.put('subscriptionId', subscription.fr_ID__c);
+        request.put('APEX_TESTS', 'true');
```
Can this parameter get removed in all cases where it was added?  As shared previously, I am worried about behavior that is specific to tests that this seems to indicate it is

gabrielcouto357 replied:
Removed!

Alexalex1432 commented on src/classes/frDonationTest.cls:
```diff
@@ -445,34 +888,91 @@ public class frDonationTest {
     }
-    
+
+    static testMethod void syncEntity_oppContactRoles_sameSupporter_NPC() {
```
It looks like the tests were not written for this method, should it be removed?

gabrielcouto357 replied:
This was just a placeholder and wasn't needed, removed!

Alexalex1432 commented on src/classes/frModel.cls:
```diff
@@ -138,11 +136,12 @@ public with sharing abstract class frModel {
         for(String fieldName : frFieldToMappings.keySet()) {
             if(request.containsKey(fieldName)) {
                 for(frMapping__c mapping : frFieldToMappings.get(fieldName)) {
+                    if(!fields.containsKey(mapping.sf_Name__c)) continue; //Make sure field is valid
```
Can we either remove this (since it shouldn't be possible unless a field was deleted after it was mapped) OR log an error in this situation where we are trying to map to a field that doesn't exist?  It might be the case that removing this handling will cause an exception to throw which will hopefully log an error without us having to do anything extra

gabrielcouto357 replied:
Removed!

Alexalex1432 commented on src/customMetadata/frField.supported.md:
As far as I am aware, there is not a field called `supported` on incoming Subscription requests.  Was there a typo here or should this be removed?

gabrielcouto357 replied:
I believe there is one but doubt it would be useful so it can removed with no issue if we wanted too!

Alexalex1432 commented on src/customMetadata/frField.frequency.md:
It looks like we only have frFields for 4 of the incoming fields on a Subscription, but there are around 28 fields synced for Subscriptions.  You can see those all in the test class here: https://github.com/funraise/Funraise-Salesforce/blob/master/src/classes/frSubscriptionTest.cls#L208

It seems like we should have frFields created for all of them so they can be mapped if organizations want to

gabrielcouto357 replied:
I think we may have discussed this previously but the thought process originally was to dataload/import these outside the package since we didnt want these fields in the package. If this has changed i can add the fields into the package, lmk what you think thanks!

gabrielcouto357 committed:
Remove continue line in model

gabrielcouto357 committed:
Update test for NPC to correctly map to person fields

gabrielcouto357 committed:
Removed unused test

gabrielcouto357 committed:
removed unused tests and apex tests references

gabrielcouto357 committed:
Uploading current fields used in dev

gabrielcouto357 committed:
Uploading the rest of the fields in the response for sub

gabrielcouto357 committed:
Update for _GC usage in file names due to limitation for duplicates when using deb name

gabrielcouto357 committed:
Delete src/customMetadata/frField.supported.md

gabrielcouto357 committed:
Delete src/customMetadata/frField.frequency.md

gabrielcouto357 committed:
Delete src/customMetadata/frField.nextPaymentDate.md

gabrielcouto357 committed:
Delete src/customMetadata/frField.pledgeAmount.md

gabrielcouto357 committed:
Delete src/customMetadata/frField.subscriptionName.md

gabrielcouto357 committed:
Added supporterId and imported to commitment fields
